newIDE/app/src/EventsFunctionsList/index.js

```diff
@@ -6,509 +6,1211 @@ import { t } from '@lingui/macro';
+        forceUpdateList,
+        onAddEventsFunction,
+        onDeleteEventsFunction,
+        onEventsFunctionAdded,
+        onRenameEventsFunction,
+        onSelectEventsFunction,
+        preferences,
+        project,
+        scrollToItem,
+        showDeleteConfirmation,
+        unsavedChanges,
+      ]
     );
 
-    this.forceUpdateList();
-  };
-
-  forceUpdateList = () => {
-    this._onEventsFunctionModified();
-    if (this.sortableList) this.sortableList.forceUpdateGrid();
-  };
-
-  _copyEventsFunction = (eventsFunction: gdEventsFunction) => {
-    Clipboard.set(EVENTS_FUNCTION_CLIPBOARD_KIND, {
-      eventsFunction: serializeToJSObject(eventsFunction),
-      name: eventsFunction.getName(),
-    });
-  };
-
-  _cutEventsFunction = (eventsFunction: gdEventsFunction) => {
-    this._copyEventsFunction(eventsFunction);
-    this._deleteEventsFunction(eventsFunction, { askForConfirmation: false });
-  };
-
-  _pasteEventsFunction = (index: number) => {
-    if (!Clipboard.has(EVENTS_FUNCTION_CLIPBOARD_KIND)) return;
-
-    const clipboardContent = Clipboard.get(EVENTS_FUNCTION_CLIPBOARD_KIND);
-    const copiedEventsFunction = SafeExtractor.extractObjectProperty(
-      clipboardContent,
-      'eventsFunction'
+    const setSelectionToFunction = React.useCallback(
+      (
+        eventsFunction: gdEventsFunction,
+        eventsBasedBehavior: ?gdEventsBasedBehavior,
+        eventsBasedObject: ?gdEventsBasedObject
+      ) => {
+        const eventsBasedEntity = eventsBasedBehavior || eventsBasedObject;
+        const eventsFunctionsContainer = eventsBasedEntity
+          ? eventsBasedEntity.getEventsFunctions()
+          : eventsFunctionsExtension;
+        const eventFunctionProps = {
+          eventsBasedObject,
+          eventsFunctionsContainer,
+          ...eventFunctionCommonProps,
+        };
+        setSelectedItems([
+          new LeafTreeViewItem(
+            new FunctionTreeViewItemContent(eventsFunction, eventFunctionProps)
+          ),
+        ]);
+        onSelectEventsFunction(
+          eventsFunction,
+          eventsBasedBehavior,
+          eventsBasedObject
+        );
+      },
+      [
+        eventFunctionCommonProps,
+        eventsFunctionsExtension,
+        onSelectEventsFunction,
+      ]
     );
-    const name = SafeExtractor.extractStringProperty(clipboardContent, 'name');
-    if (!name || !copiedEventsFunction) return;
 
-    const { project, eventsFunctionsContainer } = this.props;
+    const addNewEventsFunction = React.useCallback(
+      (itemContent: ?TreeViewItemContent) => {
+        const eventsFunctionsContainer = itemContent
+          ? itemContent.getEventsFunctionsContainer() ||
+            eventsFunctionsExtension
+          : eventsFunctionsExtension;
 
-    const newName = newNameGenerator(name, name =>
-      eventsFunctionsContainer.hasEventsFunctionNamed(name)
-    );
+        const index = itemContent && itemContent.getFunctionInsertionIndex();
```

Comment:
`getFunctionInsertionIndex` and `getEventsFunctionsContainer` are only used here. It doesn't feel like a good encapsulation. I wonder if there should rather be an `addFunction` passed in props to the `TreeViewItemContent` which would give the container and index to this function. But it would add an other callback level and may make the call tree hard to follow.